keps/sig-cluster-lifecycle/kubeadm/4656-add-instance-configuration/README.md

```diff
@@ -0,0 +1,382 @@
+* The `--container-runtime-endpoint` flag is no longer written to the `/var/lib/kubelet/kubeadm-flags.env` file
+
+### Non-Goals
+
+- Continue maintaining CRI socket paths on Node objects.
+
+## Proposal
+
+kubeadm init: If the CRI socket provided in the kubeadm configuration is set, it will take precedence and generate the `/var/lib/kubelet/instace-config.yaml` configuration file based on it; if the CRI socket is not specified, the container runtime ennpoint will be automatically detected and the configuration file will be uploaded to the global configuration.
+
+kubeadm join: If the CRI socket provided in the kubeadm configuration is set, it will take precedence and generate the `/var/lib/kubelet/instace-config.yaml` configuration file based on it, overwriting the kubelet configuration downloaded from the global configuration; if the CRI socket is not specified, the kubelet configuration downloaded from the global configuration is used, and /var/lib/kubelet/instace-config.yaml is generated based on it.
+
+kubeadm upgrade: `kubeadm upgrade node` will only check `/var/lib/kubelet/instace-config.yaml`.
+
+### Risks and Mitigations
+
+## Design Details
+
+**Alpha(v.1.31):** 
+
+kubeadm init:
+
+* No longer need to write the `container-runtime-endpoint` to `/var/lib/kubelet/kubeadm-flags.env`.
+* No longer need to add the `kubeadm.alpha.kubernetes.io/cri-socket` annotation.
+* If the CRI socket provided in the kubeadm configuration is set, it is used first and the `/var/lib/kubelet/instace-config.yaml` configuration file is generated based on it. If the CRI socket is not set, the container runtime endpoint is automatically detected and the configuration file is uploaded to the global configuration.
+
+kubeadm join:
+
+* No longer need to add the kubeadm.alpha.kubernetes.io/cri-socket annotation.
+* If the CRI socket provided in the kubeadm configuration is set, it will take precedence and generate the `/var/lib/kubelet/instace-config.yaml` configuration file based on it, overwriting the kubelet configuration downloaded from the global configuration; if the CRI socket is not set, the kubelet configuration downloaded from the global configuration is used, and /var/lib/kubelet/instace-config.yaml is generated based on it.
+
+kubeadm reset:
+
+* There is no need to do anything, as `/var/lib/kubelet/instace-config.yaml` will be cleaned up with the deletion of ``/var/lib/kubelet.`
+
+kubeadm upgrade apply:
+
+* `kubeadm upgrade apply` will read the CRI annotaion and patch it to the global configuration and generate `/var/lib/kubelet/instace-config.yaml`, and override the `ContainerRuntimeEndpoint` field to `/var/lib/kubelet/config.yaml`.
+
+kubeadm upgrade node:
+
+* for other nodes, `kubeadm upgrade node` will check if the CRI annotation and  generate `/var/lib/kubelet/instace-config.yaml` based on them, and override the `ContainerRuntimeEndpoint` field to `/var/lib/kubelet/config.yaml`.
```

Comment:
can it check the kubeadm-flags.env instance to avoid an API call?

Reply:
Yes, we can do that.